--- ../samba-samba-4.10.8/source3/libsmb/clidfs.c.old	2019-09-03 11:34:57.528817931 -0400
+++ ../samba-samba-4.10.8/source3/libsmb/clidfs.c	2019-09-03 11:10:33.552164568 -0400
@@ -259,6 +259,14 @@
 		return status;
 	}
 
+	if (*c->server_domain) {
+		DEBUG(0,("Domain=[%s] OS=[%s] Server=[%s]\n",
+			c->server_domain,c->server_os,c->server_type));
+	} else if (*c->server_os || *c->server_type) {
+		DEBUG(0,("OS=[%s] Server=[%s]\n",
+			 c->server_os,c->server_type));
+	}
+
 	DEBUG(4,(" session setup ok\n"));
 
 	/* here's the fun part....to support 'msdfs proxy' shares
